### PR TITLE
Add quantization codec and implement compression/decompression of floating point data

### PR DESCRIPTION
**Pull request by astrofrog**
Fixes https://github.com/aperiosoftware/astropy/issues/14
Fixes https://github.com/aperiosoftware/astropy/issues/40

### CODE REVIEW

**Comment by astrofrog** [comment]
I think lossy tests are close to passing but due to the issue mentioned in https://github.com/aperiosoftware/astropy/issues/40  the floating point values don't match exactly:

```
E       Mismatched elements: 144 / 144 (100%)
E       Max absolute difference: 7.677353
E       Max relative difference: 7.3442245
E        x: array([[163.9622  ,  12.696566,   2.805807,  83.07929 ,   3.645739,
E                12.367053, 118.553955,  33.24904 ,  59.632137, 209.43936 ,
E                23.05459 ,  67.77326 ],...
E        y: array([[1.626797e+02, 9.415493e+00, 1.536737e+00, 8.256937e+01,
E               5.992008e+00, 9.394189e+00, 1.165274e+02, 3.576727e+01,
E               6.091895e+01, 2.120618e+02, 2.193155e+01, 6.678799e+01],...
```

I think we should adjust the tests to compare the data read in by both libraries rather than comparing to the very original data.

**Review by Cadair — APPROVED** [review]

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 1]
```diff
@@ -647,9 +788,48 @@ def decompress_hdu(hdu):
...
+
+        if irow == 0 and hdu._header["ZCMPTYPE"] == "GZIP_2":
+            # Decompress with GZIP_1 just to find the total number of
+            # elements in the uncompressed data
+            tile_data = np.asarray(
+                decompress_tile(row["COMPRESSED_DATA"], algorithm="GZIP_1")
+            )
+            settings["itemsize"] = tile_data.size // int(np.product(actual_tile_shape))
+
+        lossless = len(cdata) == 0
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tiled_compression.py:801]: Yes that's correct

[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tiled_compression.py:801]: gzip_fallback?

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 2]
```diff
@@ -75,6 +86,121 @@ def decode(self, buf, out=None):  # pragma: no cover
             """
 
 
+class Quantize(Codec):
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tiled_compression.py:89]: That's fine and we are going to remove references to Codec anyway so we can just change this to not inherit from anything?

**Comment by astrofrog** [comment]
@Cadair - I found some time to implement the comments

**Pull request merged**